### PR TITLE
Refactor hubble redact settings schema

### PR DESCRIPTION
**Pull request by ChrsMark**
This PR implements the refactoring of the `--hubble-redact` option in order to provide more flexibility as per https://github.com/cilium/cilium/issues/23887#issuecomment-1633015416.

Specifically we change 

```yml
hubble:
 redact:
  - http-url-query
  - kafka-api-key
```

to 

```yml
hubble:
  redact:
    enabled: true
    http:
      urlQuery: true
    kafka:
      apiKey: true
```

and 

```console
--hubble-redact=http-url-query,kafka-api-key
```

to

```console
--hubble-redact=true
--hubble-redact-http-url-query=true
--hubble-redact-kafka-api-key=true
```

With this change we aim to support the option for user provided http-headers list.

This will look like the following (as per https://github.com/cilium/cilium/issues/23887#issuecomment-1632167762):

```yml
hubble:
  redact:
    enabled: false
    http:
      userInfo: true
      urlQuery: true
      headers:
        allow:
          - traceparent
          - tracestate
          - Cache-Control
          - ...
    kafka:
      apiKey: true
```

cc: @ioandr

### CODE REVIEW

File: pkg/defaults/defaults.go
Comment (pippolo84) [inline pkg/defaults/defaults.go]: Couple of typos: `HubbleReact` -> `HubbleRedact` and `is` -> `if`

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Left a comment about a couple of typos. Apart from that, agent related changes LGTM.
CI Failures are legit, you need to run `make -C Documentation/ update-cmdref` to regenerate the docs.
Thanks!

File: daemon/cmd/daemon_main.go
Comment (chancez) [inline daemon/cmd/daemon_main.go:980]: Do we even need a separate enabled option? The other options seem like they would be sufficient.

**Review by chancez — CHANGES_REQUESTED** [review]
Looks like the tests need updating. One question about if we need all 3 options now or if we could just have the 2 new ones.

File: daemon/cmd/daemon_main.go
Comment (ChrsMark) [reply daemon/cmd/daemon_main.go:980]: That's based on "this approach would provide the best flexibility for the user yet with reasonable default, all the user has to do is to set `hubble.redact.enabled=true`" from https://github.com/cilium/cilium/issues/23887#issuecomment-1632167762.

Comment (chancez) [reply daemon/cmd/daemon_main.go:980]: Ah gotcha, so the defaults for the individual options are true, but the entire feature is off by default, that makes sense to me.

Comment (ChrsMark) [reply daemon/cmd/daemon_main.go:980]: Right! Note that the defaults are still `false` in this PR, but will be fixed accordingly based on the plan I suggested at https://github.com/cilium/cilium/issues/23887#issuecomment-1646650350.

**Review by zacharysarah — APPROVED** [review]
LGTM

**Review by chancez — APPROVED** [review]
Overall LGTM.

**Review by gandro — APPROVED** [review]
Helm looks good to me. I'll defer the Hubble changes to Chance

**Review by pippolo84 — APPROVED** [review]
Thanks! :100:

**Comment by ChrsMark** [comment]
@derailed @tommyp1ckles @nathanjsweet could you please review this one when you get the chance 🙏🏽 ?

**Review by rolinh — APPROVED** [review]

**Comment by ldelossa** [comment]
All required tests are passing and have critical mass on review approvals. Merging.

**Comment by joestringer** [comment]
Both of the tests that failed on this PR prior to merging had this error, which is directly related to the changes in this PR:

```
connectivity test failed: unmarshaling cilium runtime config json: json: cannot unmarshal bool into Go struct field DaemonConfig.HubbleRedact of type []string
```

**Comment by ldelossa** [comment]
@ChrsMark as explained here: https://github.com/cilium/cilium/pull/27352#issuecomment-1670007214

this pr was merged prematurely. Can you please open a subsequent PR with any suggested fixes mentioned in this one

**Comment by ChrsMark** [comment]
Hey @ldelossa , sure I will have a look soon.
In the meantime, I wonder why we hit these compatability issues,  since https://github.com/cilium/cilium/pull/25844 should only be landed in `main` by now based on the discussions happened at https://github.com/cilium/cilium/issues/23887#issuecomment-1645784815 and the respective revert from `v1.14`.
Any pointer here would be appreciated :).

**Comment by joestringer** [comment]
@ChrsMark Ah, I didn't recognize that this is only changing code in the main branch. The reason for the incompatibility is https://github.com/cilium/cilium-cli/issues/1896 . It looks like maybe the cilium/cilium-cli repo pulled in vendor code from cilium/cilium main after https://github.com/cilium/cilium/pull/25844, so the type it internally uses is incompatible with the type that was changed in this PR.

**Comment by ChrsMark** [comment]
Hey @joestringer! So what is the best way to move this one forward? 

Given that the "breaking" change only applies on `main` would that make sense to make it now and then cilium-cli can pull an updated version of the dependency? 

Otherwise we can just rename `HubbleRedact` to `HubbleRedactEnabled` in this current patch but this would be a bit overkill just to tackle a non actual breaking change.

**Comment by joestringer** [comment]
We're in a bit of a pickle because Cilium CLI is incorrectly relying on internal Cilium structures. Either we need to solve https://github.com/cilium/cilium-cli/issues/1896, or maybe use a different name for the field in `DaemonConfig` and open a fresh PR for these changes. We might even be able to also delete the current `HubbleRedact` field. However, those changes would need to be merged into main and then pulled into cilium-cli as a dependency before cilium/cilium main could have a different field called `HubbleRedact` inside `DaemonConfig`.

cc @cilium/cilium-cli-maintainers

**Comment by michi-covalent** [comment]
yeah we still don't have a good answer for https://github.com/cilium/cilium-cli/issues/1896#issuecomment-1671605371.

in the meantime if we pick up the latest cilium code from main branch in cilium-cli, this error `connectivity test failed: unmarshaling cilium runtime config json: json: cannot unmarshal bool into Go struct field DaemonConfig.HubbleRedact of type []string` will go away right?

**Comment by michi-covalent** [comment]
... so more concretely what i'm suggesting as a short term fix is:

- revert this revert: https://github.com/cilium/cilium/pull/27352
- and then immediately pick up cilium from the latest main branch in cilium-cli repo 💭

**Comment by joestringer** [comment]
@michi-covalent We can't run CI properly and confirm that this PR doesn't break _other_ CI jobs without resolving this problem. Hence I suggest we just use a different field name so that the PR can pass the full CI and go through review *before* we merge it.

**Comment by joestringer** [comment]
> in the meantime if we pick up the latest cilium code from main branch in cilium-cli, this error connectivity test failed: unmarshaling cilium runtime config json: json: cannot unmarshal bool into Go struct field DaemonConfig.HubbleRedact of type []string will go away right?

No, the problem is that the flag was first introduced with a bool type, then this PR changes it to a string. Now that this PR is reverted, it is back to a bool type. Cilium-CLI is still incorrectly deserializing that option as a bool, so updating Cilium-CLI will not fix that problem.

**Comment by michi-covalent** [comment]
ok yeah using a different field name would work for sure if we are ok with changing the field name 👍

**Pull request merged**
